screeps-game-api/src/constants.rs

```diff
@@ -390,39 +394,53 @@ js_deserializable!(Terrain);
 /// implementations only operate on made-up integer constants. If you're ever
 /// using these impls manually, use the `__look_num_to_str` and
 /// `__look_str_to_num` JavaScript functions to convert.
+///
+/// To convert from a string representation directly, use
+/// [`FromStr`][std::str::FromStr] or [`Look::deserialize_from_str`] .
 #[repr(u32)]
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize_repr, Deserialize_repr)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize_repr, Deserialize_repr, FromStr)]
 #[doc(hidden)]
 #[serde(rename_all = "camelCase")]
 pub enum Look {
-    #[serde(rename = "creep")]
+    #[display("creep")]
     Creeps = 0,
-    #[serde(rename = "energy")]
+    #[display("energy")]
     Energy = 1,
-    #[serde(rename = "resource")]
+    #[display("resource")]
     Resources = 2,
-    #[serde(rename = "source")]
+    #[display("source")]
     Sources = 3,
-    #[serde(rename = "mineral")]
+    #[display("mineral")]
     Minerals = 4,
-    #[serde(rename = "structure")]
+    #[display("structure")]
     Structures = 5,
-    #[serde(rename = "flag")]
+    #[display("flag")]
     Flags = 6,
-    #[serde(rename = "constructionSite")]
+    #[display("constructionSite")]
     ConstructionSites = 7,
-    #[serde(rename = "nuke")]
+    #[display("nuke")]
     Nukes = 8,
-    #[serde(rename = "terrain")]
+    #[display("terrain")]
     Terrain = 9,
-    #[serde(rename = "tombstone")]
+    #[display("tombstone")]
     Tombstones = 10,
-    #[serde(rename = "powerCreep")]
+    #[display("powerCreep")]
     PowerCreeps = 11,
 }
 
 js_deserializable!(Look);
 
+impl Look {
+    /// Helper function for deserializing from a string rather than a fake
+    /// integer value.
+    pub fn deserialize_from_str<'de, D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        let s: Cow<'de, str> = Cow::deserialize(d)?;
```

Comment:
I really don't get the usage of the `Cow`? Serde doesn't allocate memory for strings? Is it because it may still refers to the initial "text" that is deserialized?

Reply:
`serde` won't do any allocations, but the deserializer implementations may or may not need to. `Cow` is just a way of encoding that this code is OK with taking either an allocated owned string, or a borrowed one.

For instance, when using `serde_json::from_str` to deserialize, strings without escape patterns can be deserialized as borrowing from the original string passed in. But in `serde_json::from_value`, the ownership of strings is passed from the value into the deserialized data, so we need to be able to accept an owned string too.

It'd function if this code just used `s: String`, but then it would also allocate slightly more than necessary in some situations.